Make sure all LEDs on SMC-Mixer go off on shutdown

SamWhited opened the pull request:
I noticed a moment ago that all of the LEDs on the SMC-Mixer don't go off when you shutdown Mixxx. The problem appears to be that the `EQRack` class is neither a component nor a component container, so the `forEachComponents` call that shuts down everything cleanly skips over it and the buttons it contains never get turned off. Changing it to a `ComponentContainer` fixes this, but also triggers a bug (I think) in QJSEngine where `this` is not defined in the callback which also had to be worked around.

SamWhited committed:
Make sure all LEDs on SMC-Mixer go off on shutdown

Signed-off-by: Sam Whited <sam@samwhited.com>

Swiftb0y commented on res/controllers/MVave-SMC-Mixer-scripts.js:
```diff
@@ -130,40 +130,42 @@ var SMCMixer;
-    class EqRack {
+    class EqRack extends components.ComponentContainer {
         constructor(index) {
+            super(undefined);
```
```suggestion
            super({});
```

Swiftb0y commented on res/controllers/MVave-SMC-Mixer-scripts.js:
```diff
@@ -130,40 +130,42 @@ var SMCMixer;
                 return function() {
-                    if (this.isLongPressed) {
-                        if (knob.inKey === this.inKey.replace("button_", "") || (this.inKey === "enabled" && this.knob.inKey === "super1")) {
+                    if (_this.isLongPressed) {
+                        if (knob.inKey === _this.inKey.replace("button_", "") || (_this.inKey === "enabled" && _this.knob.inKey === "super1")) {
```
wait, shouldn't it have been `knob` instead of `this` in the first place?

SamWhited replied:
oh yah, I guess it could have been. I forgot I redeclared it at the top (though also I have no idea why I did that, writing "this.knob" a few times isn't that much longer or harder). Fixed anyways. Thanks!

SamWhited commented:
Hmm, something else is still going wrong with this. It appears QJSEngine has some weird scope issues, I'm getting random errors now about stuff that definitely exists not existing. Working on it. One more commit incoming.

SamWhited commented:
My find/replace was a little too broad, I forgot the inner function has a different this. Have I complained yet about how much I hate javascript? *sigh*. Fixed and rebased, everything appears to be working now.

Swiftb0y commented on res/controllers/MVave-SMC-Mixer-scripts.js:
```diff
@@ -130,22 +130,30 @@ var SMCMixer;
             const btnInToggle = () => {
-                const knob = this.knob;
-                const origGroup = this.knob.group;
-                const origInKey = this.knob.inKey;
+                // "this" is undefined here due to a bug in QJSEngine, so we
```
`this` also always refers to the EqRack because this is an arrow function. You'll need to declare it as an anonymous function (so `const btnInToggle = function() {`), so the `this` can be appropriately set when this function is getting called.

Swiftb0y replied:
Then `this` should also work correctly (and always refer to the correct button instance instead of the `EqRack` instance).

SamWhited replied:
You'd be right except for the bug in QJSEngine I mentioned, if I call super above, this stops being defined entirely and using "this" results in an error. It definitely shouldn't be this way, but it is.

Swiftb0y replied:
Oh, no sorry for the bogus comment. I totally missed where we already return a function in the `btnInToggle` case. Yeah in that case the local alias is probably best.

Swiftb0y reviewed (approved):
LGTM. Thank you.
